### PR TITLE
Clarified condition coverage

### PR DESCRIPTION
**Pull request by djosh34**
Dear SQT staff,

For me the chapter about basic/full/branch+decision coverage was very confusing. I just couldn't get why some things were calculated one way here and another way there. However, when I've discover one little detail I think it just clicked for me. It was regarding the difference between lazy-operators and non lazy-operators. This is why I didn't get that the condition coverage in the explanation was 100%, as I just thought that you could always split up if-statements into multiple blocks.

I hope that these changes makes it clear for everyone and please make sure everything is correct as I'm still not completely sure if this is the correct explanation. 

Thanks in advance for reading my pull request,

Joshua Azimullah

### CODE REVIEW

**Comment by mauricioaniche** [comment]
Hi, @djosh34! Thanks for the feedback!

I'll read your PR carefully!

Meanwhile, if you still have questions about their differences, send me an e-mail and we schedule a Zoom call, ok?

Cheers!

**Comment by mauricioaniche** [comment]
Hi, @djosh34! I carefully read your PR!

I think you have a small misconception in your definition.

The problem with basic condition coverage is not really related to lazy operators. It's related with the fact that you might exercise conditions as true and false, and yet, just exercise one branch of the overall decision.

See the code below:

```java
if(a || b) {
  print("in");
} else {
  print("out");
}
``` 

Test cases T1=(true, false) and T2=(false, true), both print "in". They achieve 100% basic coverage, but just 50% branch coverage. The else branch was never exercised. That's bad! That's the problem when you focus only on conditions.

And that's why we suggest condition + branch coverage. In other words, as a tester, you make sure you exercise all the conditions as true and false at least once, and the overall decision block (the one containing the entire expression) as true and false at least once.

Did you get it?

If you are still in doubt about this, join one of the Q&A sessions I am setting up this week, and we talk in person over there! 

Cheers!

**Comment by djosh34** [comment]
No, I still don't really get it. This comes from the fact that I don't get why we ever would do basic condition coverage or condition+branch coverage, when there is a much more simpler way:

Take this bit of code
```
if(a || b) {
  print("in");
} else {
  print("out");
}
```

Create this kind of branch/condition CFG

Blank Diagram (3)

And then count all the branches that have been exercised using T1=(true, false) and T2=(false, true), which would equate to 3/4 = 75%

Does this way of coverage also have a name? Why don't we use this kind of coverage?

This is what I initially thought condition coverage was after watching your video on the topic. However, when reading I thought that basic condition coverage was like this:

```
if(a || b) {

  print("in");
} else {

  if(c) {
    print("inside branch");
  } 

  print("out");
}
```

And then with two tests T1=(true,true,true) and T2=(false,false,false) you had 100% basic condition coverage, which made me wonder even more why anyone would use basic condition coverage.


After a lot of re-reading and looking at the discussions on brightspace I thought that no, basic condition coverage is not what I described here above (using two tests 100%), but that it depends on which conditions you cover.

But now I'm even more confused.

```
if(a || b) {
  print("in");
} else {
  print("out");
}
```
In this bit of code (b = false) is never exercised in T1=(true,false), as in Java making a=true means that b is never exercised. This believe was strengthened due to exercise 9:
image

The answers clearly state that condition 2 isn't exercised with T2=8:
image
This makes sense to me as `&&` is lazy and only exercises the conditions that are needed.



Why is in this bit of code
```
if(a || b) {
  print("in");
} else {
  print("out");
}
```
With test T1(true,false), b = false is exercised, the basic condition coverage is 100%

But with T2=8 
image
Condition 2: Line 1 (n % 5 == 0) = false, isn't exercised




My final guess was that basic condition coverage was what you've just explained above in your comment and that full condition coverage was my way on doing it getting 75%, and finally branch+condition coverage was this formula:
image
with the conditions being the conditions exercise using basic condition coverage.

Nevertheless, you stated in the book that "full condition coverage" is the same as "condition+branch coverage", so my final guess is also not true.


To recap:

Why don't we use the branch/condition coverage tactic stated above getting 75% coverage from your example instead of branch+condition coverage?
When are conditions (with basic condition coverage) in a lazy-operator setting exercised? Is it like your example above or like exercise 9 in the book?
















I would love to join the Q&A

**Comment by mauricioaniche** [comment]
Hi, @djosh34! Can you send me an email and we schedule a Zoom conversation? It'll get easier if we talk!! Cheers!

**Comment by mauricioaniche** [comment]
Hi, @djosh34! I just had time to read your last message. Quick edit on my previous code: I should use `|` instead of `||` for my example to make sense.

I think the explanation is simpler than what you have in your mind now. Let's do it. 

We first discussed branch coverage (also known as decision coverage). There, we can about covering the entire set of conditions. So, it doesn't matter how big it is, e.g., `if(a>10)` or `if(a>10 && b < 20 || c>10...)`, we just care about making the entire if, true or false. 

If we want to be even more rigorous, we can start to think of exercising each of the conditions individually. The natural way of thinking here is: _"let's make each condition true and false"_. In this case, you achieve 100% coverage if all conditions were true and false at least once.

However, if you just follow this _"let's make each condition true and false"_ in a naive way, you might end up covering all the conditions as true and false, but not all the decision points. The simplest way to demonstrate is with examples like `a | b`. If T1=(true, false) and T2=(false, true), `a` was true and false, `b` was true and false. But the outcome was only true. 

This "naive way" of doing it is what we call _basic_ condition coverage. It's, of course, very naive, so nobody does (or should do) that. However, if we want do define/formalise a systematic criteria that takes each condition individually, we should think of:

1. Exercising each condition as true and false
1. Exercising the overall decision block as true and false

What seems to me to be confusing you is the lazy operator and how to calculate coverage. The  use of a lazy operator, e.g., `a && b` reduces the number of test cases we can come up with.

T1 a = true, b = true, outcome=true
T2 a = true, b = false, outcome=false
T3 a = false, b = doesn't matter, outcome=false

are the possibilities.

If we want to achieve full condition+branch coverage in this case, we need these three tests. T1 makes a=true, b=true. Overall decision: true. So, we are missing a=false, b=false, outcome false. T2 gives us a = true, b = false, outcome false. Now, we only miss a = false. That's what we get with T3.

In your `a || b` example, the one you have the picture. T1 and T2 exercise both a and b as true and false. But the only decision block that is exercised is the print("in"). So, it's clear we are missing a test here, right?!!! So, if we were guided only by the individual conditions, we could have said "yeah, I covered all the conditions individually", but we missed part of the program.

I also noticed that you are counting the number of edges as a way to calculate coverage. Edges covered / Total edges. That's one way to do it, but not the one I teach in the course. In here, I prefer to go for "number of decisions + number of conditions". This one doesn't require you to draw a CFG to calculate. You can do it by simply looking at the code. However, note that in both formulas, yours and mine, if you cover all conditions AND all branches, both formulas will give you 100%. 

Does that clarify things?
